File: discipline_ts/core/chronic/time.ts
import { Duration, TextualError, Nominal, Tried, FailureCode } from "../x.ts";

const brand = Symbol();

export type Time = Nominal<typeof brand, number>;

export const construct = (timestamp: number): Time => {
  return Nominal.create(brand, timestamp);
};

export const MINIMUM_TIMESTAMP = 0;
export const MAXIMUM_TIMESTAMP = 1000 * 60 * 60 * 24 - 1;

export const fromTimestamp = (timestamp: number): Tried<Time, TextualError> => {
  if (!Number.isInteger(timestamp)) {
    const it = TextualError.create("Creating a Time from a millisecond timestamp since midnight");
    TextualError.addMessage(it, "Argument 'timestamp' is not an integer");
    TextualError.addNumberAttachment(it, "Argument 'timestamp'", timestamp);
    return Tried.Failure(it);
  }

  if (timestamp < MINIMUM_TIMESTAMP) {
    const it = TextualError.create("Creating a Time from a millisecond timestamp since midnight");
    TextualError.addMessage(it, "Argument 'timestamp' is less than the minimum valid value");
    TextualError.addNumberAttachment(it, "Argument 'timestamp'", timestamp);
    TextualError.addNumberAttachment(it, "Minimum valid value", MINIMUM_TIMESTAMP);
    return Tried.Failure(it);
  }

  if (timestamp > MINIMUM_TIMESTAMP) {
    const it = TextualError.create("Creating a Time from a millisecond timestamp since midnight");
    TextualError.addMessage(it, "Argument 'timestamp' is greater than the maximum valid value");
    TextualError.addNumberAttachment(it, "Argument 'timestamp'", timestamp);
    TextualError.addNumberAttachment(it, "Maximum valid value", MAXIMUM_TIMESTAMP);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(timestamp));
};

export const fromTimestampOrError = (
  timestamp: number,
  textualError: TextualError,
): Time | FailureCode => {

};

export const fromHourAndMinuteAm = (hour: number, minute: number): Tried<Time, TextualError> => {
  if (!Number.isInteger(hour)) {
    const it = TextualError.create("Creating a Time from a hour (AM) and minute arguments");
    TextualError.addMessage(it, "Argument 'hour' is not an integer");
    TextualError.addNumberAttachment(it, "Argument 'hour'", hour);
    return Tried.Failure(it);
  }
  if (hour < 0) {
    const it = TextualError.create("Creating a Time from a hour (AM) and minute arguments");
    TextualError.addMessage(it, "Argument 'hour' is less than '0'");
    TextualError.addNumberAttachment(it, "Argument 'hour'", hour);
    return Tried.Failure(it);
  }
  if (hour > 11) {
    const it = TextualError.create("Creating a Time from a hour (AM) and minute arguments");
    TextualError.addMessage(it, "Argument 'hour' is greater than '11'");
    TextualError.addNumberAttachment(it, "Argument 'hour'", hour);
    return Tried.Failure(it);
  }

  if (!Number.isInteger(minute)) {
    const it = TextualError.create("Creating a Time from a minute (AM) and minute arguments");
    TextualError.addMessage(it, "Argument 'minute' is not an integer");
    TextualError.addNumberAttachment(it, "Argument 'minute'", minute);
    return Tried.Failure(it);
  }
  if (minute < 0) {
    const it = TextualError.create("Creating a Time from a minute (AM) and minute arguments");
    TextualError.addMessage(it, "Argument 'minute' is less than '0'");
    TextualError.addNumberAttachment(it, "Argument 'minute'", minute);
    return Tried.Failure(it);
  }
  if (minute > 59) {
    const it = TextualError.create("Creating a Time from a minute (AM) and minute arguments");
    TextualError.addMessage(it, "Argument 'minute' is greater than '59'");
    TextualError.addNumberAttachment(it, "Argument 'minute'", minute);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(
    hour * Duration.MILLISECONDS_PER_HOUR
    +
    minute * Duration.MILLISECONDS_PER_MINUTE
  ));
};

export const fromHourAndMinutePm = (hour: number, minute: number): Tried<Time, TextualError> => {
  if (!Number.isInteger(hour)) {
    const it = TextualError.create("Creating a Time from a hour (PM) and minute arguments");
    TextualError.addMessage(it, "Argument 'hour' is not an integer");
    TextualError.addNumberAttachment(it, "Argument 'hour'", hour);
    return Tried.Failure(it);
  }
  if (hour < 0) {
    const it = TextualError.create("Creating a Time from a hour (PM) and minute arguments");
    TextualError.addMessage(it, "Argument 'hour' is less than '0'");
    TextualError.addNumberAttachment(it, "Argument 'hour'", hour);
    return Tried.Failure(it);
  }
  if (hour > 11) {
    const it = TextualError.create("Creating a Time from a hour (PM) and minute arguments");
    TextualError.addMessage(it, "Argument 'hour' is greater than '11'");
    TextualError.addNumberAttachment(it, "Argument 'hour'", hour);
    return Tried.Failure(it);
  }

  if (!Number.isInteger(minute)) {
    const it = TextualError.create("Creating a Time from a minute (PM) and minute arguments");
    TextualError.addMessage(it, "Argument 'minute' is not an integer");
    TextualError.addNumberAttachment(it, "Argument 'minute'", minute);
    return Tried.Failure(it);
  }
  if (minute < 0) {
    const it = TextualError.create("Creating a Time from a minute (PM) and minute arguments");
    TextualError.addMessage(it, "Argument 'minute' is less than '0'");
    TextualError.addNumberAttachment(it, "Argument 'minute'", minute);
    return Tried.Failure(it);
  }
  if (minute > 59) {
    const it = TextualError.create("Creating a Time from a minute (PM) and minute arguments");
    TextualError.addMessage(it, "Argument 'minute' is greater than '59'");
    TextualError.addNumberAttachment(it, "Argument 'minute'", minute);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(
    (12 + hour) * Duration.MILLISECONDS_PER_HOUR
    +
    minute * Duration.MILLISECONDS_PER_MINUTE
  ));
};

export const fromHourAndMinute = (hour: number, minute: number): Tried<Time, TextualError> => {
  if (!Number.isInteger(hour)) {
    const it = TextualError.create("Creating a Time from a hour and minute arguments");
    TextualError.addMessage(it, "Argument 'hour' is not an integer");
    TextualError.addNumberAttachment(it, "Argument 'hour'", hour);
    return Tried.Failure(it);
  }
  if (hour < 0) {
    const it = TextualError.create("Creating a Time from a hour and minute arguments");
    TextualError.addMessage(it, "Argument 'hour' is less than '0'");
    TextualError.addNumberAttachment(it, "Argument 'hour'", hour);
    return Tried.Failure(it);
  }
  if (hour > 23) {
    const it = TextualError.create("Creating a Time from a hour and minute arguments");
    TextualError.addMessage(it, "Argument 'hour' is greater than '23'");
    TextualError.addNumberAttachment(it, "Argument 'hour'", hour);
    return Tried.Failure(it);
  }

  if (!Number.isInteger(minute)) {
    const it = TextualError.create("Creating a Time from a minute and minute arguments");
    TextualError.addMessage(it, "Argument 'minute' is not an integer");
    TextualError.addNumberAttachment(it, "Argument 'minute'", minute);
    return Tried.Failure(it);
  }
  if (minute < 0) {
    const it = TextualError.create("Creating a Time from a minute and minute arguments");
    TextualError.addMessage(it, "Argument 'minute' is less than '0'");
    TextualError.addNumberAttachment(it, "Argument 'minute'", minute);
    return Tried.Failure(it);
  }
  if (minute > 59) {
    const it = TextualError.create("Creating a Time from a minute and minute arguments");
    TextualError.addMessage(it, "Argument 'minute' is greater than '59'");
    TextualError.addNumberAttachment(it, "Argument 'minute'", minute);
    return Tried.Failure(it);
  }

  return Tried.Success(construct(
    hour * Duration.MILLISECONDS_PER_HOUR
    +
    minute * Duration.MILLISECONDS_PER_MINUTE
  ));
};

export const getTimestamp = (it: Time): number => {
  return Nominal.get(it);
};

export const getHour = (it: Time): number => {
  return Math.floor(
    getTimestamp(it) 
    /
    Duration.MILLISECONDS_PER_HOUR
  );
};

export const getMinute = (it: Time): number => {
  return Math.floor(
    getTimestamp(it) 
    % 
    Duration.MILLISECONDS_PER_HOUR 
    / 
    Duration.MILLISECONDS_PER_MINUTE
  );
};

export const getSecond = (it: Time): number => {
  return Math.floor(
    getTimestamp(it)
    % 
    Duration.MILLISECONDS_PER_HOUR 
    %
    Duration.MILLISECONDS_PER_MINUTE
    /
    Duration.MILLISECONDS_PER_SECOND
  );
};

export const toString = (it: Time): string => {
  return `${
    getHour(it).toString()
  }:${
    getMinute(it).toString()
  }:${
    getSecond(it).toString()
  }`;
};

export const Time = {
  MINIMUM_TIMESTAMP,
  MAXIMUM_TIMESTAMP,
  fromTimestamp,
  fromTimestampOrError,
  fromHourAndMinute,
  fromHourAndMinuteAm,
  fromHourAndMinutePm,
  getTimestamp,
  getHour,
  getMinute,
  getSecond,
  toString,
};

File: discipline4/app/src/main/java/com/example/app/chronic/Time.kt
package com.example.app

import com.example.app.TextualError
import com.example.app.Duration
import androidx.room.Entity

/**
 * Represents a time of day (no date component) with millisecond precision
 */
@Entity
@JvmInline
value class Time private constructor(
  private val timestamp: Int
) {
  
  companion object {
    // 00:00:00.000
    const val MINIMUM_TIMESTAMP = 0

    // 23:59:59.999
    const val MAXIMUM_TIMESTAMP = 1000 * 60 * 60 * 24 - 1  
    
    private const val MILLISECONDS_PER_HOUR = Duration.MILLISECONDS_PER_HOUR.toInt()
    private const val MILLISECONDS_PER_MINUTE = Duration.MILLISECONDS_PER_MINUTE.toInt()
    private const val MILLISECONDS_PER_SECOND = Duration.MILLISECONDS_PER_SECOND.toInt()
    
    /**
     * Creates a Time from a timestamp (milliseconds since midnight)
     */
    fun fromTimestamp(timestamp: Int): Tried<Time, TextualError> {
      if (timestamp < MINIMUM_TIMESTAMP) {
        return Tried.failure(
          TextualError
            .create("Creating a Time from a millisecond timestamp since midnight")
            .addMessage("Argument 'timestamp' is less than the minimum valid value")
            .addIntAttachment("Argument 'timestamp'", timestamp)
            .addIntAttachment("Minimum valid value", MINIMUM_TIMESTAMP)
        )
      }
      
      if (timestamp > MAXIMUM_TIMESTAMP) {
        return Tried.failure(
          TextualError
            .create("Creating a Time from a millisecond timestamp since midnight")
            .addMessage("Argument 'timestamp' is greater than the maximum valid value")
            .addIntAttachment("Argument 'timestamp'", timestamp)
            .addIntAttachment("Maximum valid value", MAXIMUM_TIMESTAMP)
        )
      }
      
      return Tried.success(Time(timestamp))
    }

    fun fromTimestampOrThrow(timestamp: Int): Time {
      if (timestamp < MINIMUM_TIMESTAMP) {
        throw TextualError
          .create("Creating a Time from a millisecond timestamp since midnight")
          .addMessage("Argument 'timestamp' is less than the minimum valid value")
          .addIntAttachment("Argument 'timestamp'", timestamp)
          .addIntAttachment("Minimum valid value", MINIMUM_TIMESTAMP)
      }

      if (timestamp > MAXIMUM_TIMESTAMP) {
        throw TextualError
          .create("Creating a Time from a millisecond timestamp since midnight")
          .addMessage("Argument 'timestamp' is greater than the maximum valid value")
          .addIntAttachment("Argument 'timestamp'", timestamp)
          .addIntAttachment("Maximum valid value", MAXIMUM_TIMESTAMP)
      }

      return Time(timestamp)
    }
    
    /**
     * Creates a Time from hour and minute (24-hour format)
     */
    fun fromHourAndMinute(hour: Int, minute: Int): Tried<Time, TextualError> {
      if (hour !in 0..23) {
        return Tried.failure(
          TextualError
            .create("Creating a Time from hour and minute arguments")
            .addMessage("Argument 'hour' must be between 0 and 23")
            .addIntAttachment("Argument 'hour'", hour)
        )
      }
      
      if (minute !in 0..59) {
        return Tried.failure(
          TextualError
            .create("Creating a Time from hour and minute arguments")
            .addMessage("Argument 'minute' must be between 0 and 59")
            .addIntAttachment("Argument 'minute'", minute)
        )
      }
      
      return Tried.success(
        Time(hour * MILLISECONDS_PER_HOUR + minute * MILLISECONDS_PER_MINUTE)
      )
    }

    fun fromHourAndMinuteOrThrow(hour: Int, minute: Int): Time {
      if (hour !in 0..23) {
        throw TextualError
          .create("Creating a Time from hour and minute arguments")
          .addMessage("Argument 'hour' must be between 0 and 23")
          .addIntAttachment("Argument 'hour'", hour)
      }

      if (minute !in 0..59) {
        throw TextualError
          .create("Creating a Time from hour and minute arguments")
          .addMessage("Argument 'minute' must be between 0 and 59")
          .addIntAttachment("Argument 'minute'", minute)
      }

      return Time(hour * MILLISECONDS_PER_HOUR + minute * MILLISECONDS_PER_MINUTE)
    }
    
    /**
     * Creates a Time from hour and minute (12-hour AM format)
     */
    fun fromHourAndMinuteAm(hour: Int, minute: Int): Tried<Time, TextualError> {
      if (hour !in 0..11) {
        return Tried.failure(
          TextualError
            .create("Creating a Time from hour (AM) and minute arguments")
            .addMessage("Argument 'hour' must be between 0 and 11")
            .addIntAttachment("Argument 'hour'", hour)
        )
      }
      
      if (minute !in 0..59) {
        return Tried.failure(
          TextualError
            .create("Creating a Time from hour (AM) and minute arguments")
            .addMessage("Argument 'minute' must be between 0 and 59")
            .addIntAttachment("Argument 'minute'", minute)
        )
      }
      
      return Tried.success(
        Time(hour * MILLISECONDS_PER_HOUR + minute * MILLISECONDS_PER_MINUTE)
      )
    }

    fun fromHourAndMinuteAmOrThrow(hour: Int, minute: Int): Time {
      if (hour !in 0..11) {
        throw TextualError
          .create("Creating a Time from hour (AM) and minute arguments")
          .addMessage("Argument 'hour' must be between 0 and 11")
          .addIntAttachment("Argument 'hour'", hour)
      }

      if (minute !in 0..59) {
        throw TextualError
          .create("Creating a Time from hour (AM) and minute arguments")
          .addMessage("Argument 'minute' must be between 0 and 59")
          .addIntAttachment("Argument 'minute'", minute)
      }

      return Time(hour * MILLISECONDS_PER_HOUR + minute * MILLISECONDS_PER_MINUTE)
    }
    
    /**
     * Creates a Time from hour and minute (12-hour PM format)
     */
    fun fromHourAndMinutePm(hour: Int, minute: Int): Tried<Time, TextualError> {
      if (hour !in 0..11) {
        return Tried.failure(
          TextualError
            .create("Creating a Time from hour (PM) and minute arguments")
            .addMessage("Argument 'hour' must be between 0 and 11")
            .addIntAttachment("Argument 'hour'", hour)
        )
      }
      
      if (minute !in 0..59) {
        return Tried.failure(
          TextualError
            .create("Creating a Time from hour (PM) and minute arguments")
            .addMessage("Argument 'minute' must be between 0 and 59")
            .addIntAttachment("Argument 'minute'", minute)
        )
      }
      
      return Tried.success(
        Time((12 + hour) * MILLISECONDS_PER_HOUR + minute * MILLISECONDS_PER_MINUTE)
      )
    }

    fun fromHourAndMinutePmOrThrow(hour: Int, minute: Int): Time {
      if (hour !in 0..11) {
        throw TextualError
          .create("Creating a Time from hour (PM) and minute arguments")
          .addMessage("Argument 'hour' must be between 0 and 11")
          .addIntAttachment("Argument 'hour'", hour)
      }

      if (minute !in 0..59) {
        throw TextualError
          .create("Creating a Time from hour (PM) and minute arguments")
          .addMessage("Argument 'minute' must be between 0 and 59")
          .addIntAttachment("Argument 'minute'", minute)
      }

      return Time((12 + hour) * MILLISECONDS_PER_HOUR + minute * MILLISECONDS_PER_MINUTE)
    }
  }
  
  fun toTimestamp(): Int {
    return timestamp
  }
  
  fun getHour(): Int {
    return timestamp / MILLISECONDS_PER_HOUR
  }
  
  fun getMinute(): Int {
    return (timestamp % MILLISECONDS_PER_HOUR) / MILLISECONDS_PER_MINUTE
  }
  
  fun getSecond(): Int {
    return (timestamp % MILLISECONDS_PER_MINUTE) / MILLISECONDS_PER_SECOND
  }
  
  fun getMillisecond(): Int {
    return timestamp % MILLISECONDS_PER_SECOND
  }  
  
  /* BE CRAEFULL: THIS MIGHT THROW */
  override fun toString(): String {
    return String.format("%02d:%02d:%02d", getHour(), getMinute(), getSecond())
  }
  
  fun to12HourStringOrThrow(): String {
    val hour24 = getHour()
    val hour12 = when (hour24) {
      0 -> 12
      in 1..11 -> hour24
      in 12..23 -> hour24 - 12
      else -> 0
    }
    val period = if (hour24 < 12) "AM" else "PM"
    return String.format("%02d:%02d %s", hour12, getMinute(), period)
  }  
}
